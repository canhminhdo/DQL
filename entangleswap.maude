load BDQL.maude

fmod ENTAGLE-SWAP is
    pr BDQL-SEMANTICS .

    var S : Scalar .

    --- defining initial states
    ops a b : -> Scalar .
    eq Pow(Abs(a), 2) .+ Pow(Abs(b), 2) = 1 .
    eq S .* Pow(Abs(a), 2) .+ S .* Pow(Abs(b), 2) = S .

    ops EPR : -> Matrix .
    eq EPR = (1 ./ Sqrt(2)) . (|0>(x)|0> + |1>(x)|1>) .

    op initQState : -> Qubit .
    eq initQState = (q[0]: |0>) (q[1]: |0>) (q[2]: |0>) (q[3]: |0>) .
endfm

in debug.maude

--- red in ENTAGLE-SWAP : initQState |= [
--- H(1) ; CX(1, 0) ; H(3) ; CX(3, 2) ; CX(0, 3) ; H(0) ;
--- if P(3, |0>) then skip else X(2) fi ;
--- if P(0, |0>) then skip else Z(2) fi
--- ] P(1, 2, EPR) .

red in ENTAGLE-SWAP : initQState |= [
H(0) ; CX(0, 1) ; H(2) ; CX(2, 3) ; CX(1, 2) ; H(1) ;
if P(2, |0>) then skip else X(3) fi ;
if P(1, |0>) then skip else Z(3) fi ;
SWAP(1, 3)
] P(1, 0, EPR) .
