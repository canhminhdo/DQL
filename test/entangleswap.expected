==========================================
reduce in ENTAGLE-SWAP : initQState |= [H(1) ; CX(1,0) ; H(3) ; CX(3,2) ; CX(0,
    3) ; H(0) ; if P(3,|0>) then skip else X(2) fi ; if P(0,|0>) then skip else
    Z(2) fi]P(1,2,EPR) .
result JudgementSet: emptyJS
Bye.
